pkg/providers/openstack/stage.go

```diff
@@ -0,0 +1,201 @@
+		RunSpec:        params.RawImage.RunSpec,
+		StageSpec:      params.RawImage.StageSpec,
+		SizeMb:         int64(imageSizeMB),
+		Infrastructure: types.Infrastructure_OPENSTACK,
+		Created:        time.Now(),
+	}
+
+	// Update state.
+	if err := p.state.ModifyImages(func(images map[string]*types.Image) error {
+		images[createdImage.ID] = image
+		return nil
+	}); err != nil {
+		return nil, errors.New("failed to modify image map in state", err)
+	}
+
+	logrus.WithFields(logrus.Fields{"image": image}).Infof("image created succesfully")
+	return image, nil
+}
+
+// pickFlavor picks flavor that best matches criteria (i.e. HDD size and RAM size).
+// While diskMB is required, memoryMB is optional (set to -1 to ignore).
+func pickFlavor(clientNova *gophercloud.ServiceClient, diskMB int, memoryMB int) (*flavors.Flavor, error) {
+	if diskMB <= 0 {
+		return nil, errors.New("Please specify disk size.", nil)
+	}
+
+	var flavs []flavors.Flavor = listFlavors(clientNova, int(math.Ceil(float64(diskMB)/1024)), memoryMB)
+
+	// Find smallest flavor for given conditions.
+	logrus.Infof("Find smallest flavor for conditions: diskMB >= %d AND memoryMB >= %d\n", diskMB, memoryMB)
+
+	var bestFlavor flavors.Flavor
+	var minDiffDisk int = -1
+	var minDiffMem int = -1
+	for _, f := range flavs {
+		diffDisk := f.Disk*1024 - diskMB
+		var diffMem int = 0 // 0 is best value
+		if memoryMB > 0 {
+			diffMem = f.RAM - memoryMB
+		}
+
+		if diffDisk >= 0 && // disk is big enough
+			(minDiffDisk == -1 || minDiffDisk > diffDisk) && // disk is smaller than current best, but still big enough
+			diffMem >= 0 && // memory is big enough
+			(minDiffMem == -1 || minDiffMem >= diffMem) { // memory is smaller than current best, but still big enough
+			bestFlavor, minDiffDisk, minDiffMem = f, diffDisk, diffMem
+		}
+	}
+	if minDiffDisk == -1 {
+		return nil, errors.New(fmt.Sprintf("No flavor fits required conditions: diskMB >= %d AND memoryMB >= %d\n", diskMB, memoryMB), nil)
+	}
+	return &bestFlavor, nil
+}
+
+// listFlavors returns list of all flavors.
+func listFlavors(clientNova *gophercloud.ServiceClient, minDiskGB int, minMemoryMB int) []flavors.Flavor {
+	var flavs []flavors.Flavor = make([]flavors.Flavor, 0)
+
+	pagerFlavors := flavors.ListDetail(clientNova, flavors.ListOpts{
+		MinDisk: minDiskGB,
+		MinRAM:  minMemoryMB,
+	})
+	pagerFlavors.EachPage(func(page pagination.Page) (bool, error) {
+		flavorList, _ := flavors.ExtractFlavors(page)
+
+		for _, f := range flavorList {
+			flavs = append(flavs, f)
+		}
+
+		return true, nil
+	})
+	return flavs
+}
+
+// pushImage first creates meta for image at OpenStack, then it sends binary data for it, the qcow2 image.
+func pushImage(clientGlance *gophercloud.ServiceClient, imageName string, imageFilepath string, flavor *flavors.Flavor) (*images.Image, error) {
+	// Create metadata (on OpenStack).
+	createdImage, err := createImage(clientGlance, imageName, flavor)
+	if err != nil {
+		return nil, err
```

Comment:
we like to use error wrapping to clarify the stack trace & provide some kind of logical reasoning about where errors are coming from, in part because Go doesn't provide stack traces. i recommend using the pkg/errors package (used elsewhere in this file) to wrap errors like this. it's not super necessary for boilerplate error checking (e.g. deserializing json) but for calling 3rd party apis like this it's definitely useful